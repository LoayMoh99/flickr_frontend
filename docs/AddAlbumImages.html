<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Namespace: AddAlbumImages</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Namespace: AddAlbumImages</h1>

    




<section>

<header>
    
        <h2>AddAlbumImages</h2>
        
    
</header>

<article>
    <div class="container-overview">
    
        
            <div class="description"><p>Renders AddAlbumImages component  with album cover photo, title and number of existant photos</p></div>
        

        


    <h5 class="subsection-title">Properties:</h5>

    

<table class="props">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>props.albumCover</code></td>
            

            <td class="type">
            
                
<span class="param-type">String</span>


            
            </td>

            

            

            <td class="description last"><p>Album cover photo url</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>props.albumTitle</code></td>
            

            <td class="type">
            
                
<span class="param-type">String</span>


            
            </td>

            

            

            <td class="description last"><p>Album title</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>props.albumPhotos</code></td>
            

            <td class="type">
            
                
<span class="param-type">Array</span>


            
            </td>

            

            

            <td class="description last"><p>Album photos</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>props.imgAddIds</code></td>
            

            <td class="type">
            
                
<span class="param-type">Array</span>


            
            </td>

            

            

            <td class="description last"><p>Array of photos to be added to Album</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>props.albumId</code></td>
            

            <td class="type">
            
                
<span class="param-type">String</span>


            
            </td>

            

            

            <td class="description last"><p>Album id</p></td>
        </tr>

    
    </tbody>
</table>




<dl class="details">

    

    

    

    

    

    

    

    

    
    <dt class="tag-author">Author:</dt>
    <dd class="tag-author">
        <ul>
            <li>Khadija Khaled</li>
        </ul>
    </dd>
    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="CameraRoll_AddAlbumImages.jsx.html">CameraRoll/AddAlbumImages.jsx</a>, <a href="CameraRoll_AddAlbumImages.jsx.html#line7">line 7</a>
    </li></ul></dd>
    

    

    

    
</dl>


        
    
    </div>

    
        <h3 class="subsection-title">Extends</h3>

        


    <ul>
        <li>Component</li>
    </ul>


    

    

    

    

    

    

    

    
        <h3 class="subsection-title">Methods</h3>

        
            

    

    
    <h4 class="name" id=".containsPhotos"><span class="type-signature">(static) </span>containsPhotos<span class="signature">(obj)</span><span class="type-signature"> &rarr; {boolean}</span></h4>
    

    



<div class="description">
    <p>To check if it is present in the array of photos selected or not</p>
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>obj</code></td>
            

            <td class="type">
            
                
<span class="param-type">object</span>


            
            </td>

            

            

            <td class="description last"><p>A photo object</p></td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="CameraRoll_AddAlbumImages.jsx.html">CameraRoll/AddAlbumImages.jsx</a>, <a href="CameraRoll_AddAlbumImages.jsx.html#line53">line 53</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        
<div class="param-desc">
    <ul>
<li>A boolean indicating the photo id is present within the array or not</li>
</ul>
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">boolean</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id=".isPhotoPresent"><span class="type-signature">(static) </span>isPhotoPresent<span class="signature">()</span><span class="type-signature"> &rarr; {boolean}</span></h4>
    

    



<div class="description">
    <p>To check whether the selected photos ids are all present in a specific album or not</p>
</div>













<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="CameraRoll_AddAlbumImages.jsx.html">CameraRoll/AddAlbumImages.jsx</a>, <a href="CameraRoll_AddAlbumImages.jsx.html#line63">line 63</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        
<div class="param-desc">
    <ul>
<li>A boolean indicating whether teh array of selected photos is totally included inside the album phtos' array</li>
</ul>
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">boolean</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id=".toggleChecked"><span class="type-signature">(static) </span>toggleChecked<span class="signature">()</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    <p>Toggle isChecked boolean</p>
</div>













<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="CameraRoll_AddAlbumImages.jsx.html">CameraRoll/AddAlbumImages.jsx</a>, <a href="CameraRoll_AddAlbumImages.jsx.html#line30">line 30</a>
    </li></ul></dd>
    

    

    

    
</dl>




















        
    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Namespaces</h3><ul><li><a href="AddAlbumImages.html">AddAlbumImages</a></li><li><a href="AddModal.html">AddModal</a></li><li><a href="AddPhotos.html">AddPhotos</a></li><li><a href="AlbumCard.html">AlbumCard</a></li><li><a href="AlbumCoverPhoto.html">AlbumCoverPhoto</a></li><li><a href="AlbumPage.html">AlbumPage</a></li><li><a href="AlbumPreview.html">AlbumPreview</a></li><li><a href="CamreRoll.html">CamreRoll</a></li><li><a href="Card.html">Card</a></li><li><a href="CommentBox.html">CommentBox</a></li><li><a href="CreateNewAlbumModal.html">CreateNewAlbumModal</a></li><li><a href="DropdownPrivacy.html">DropdownPrivacy</a></li><li><a href="EditInfo.html">EditInfo</a></li><li><a href="EditModal.html">EditModal</a></li><li><a href="Follow.html">Follow</a></li><li><a href="Followers.html">Followers</a></li><li><a href="FollwingFollowers.html">FollwingFollowers</a></li><li><a href="GroupCard.html">GroupCard</a></li><li><a href="GroupHeader.html">GroupHeader</a></li><li><a href="GroupPhotos.html">GroupPhotos</a></li><li><a href="ImageDetails.html">ImageDetails</a></li><li><a href="ImageGrid.html">ImageGrid</a></li><li><a href="ImagesCR.html">ImagesCR</a></li><li><a href="ImagesMM.html">ImagesMM</a></li><li><a href="Join.html">Join</a></li><li><a href="Modal.html">Modal</a></li><li><a href="NavBar.html">NavBar</a></li><li><a href="PeopleCard.html">PeopleCard</a></li><li><a href="PhotosToAdd.html">PhotosToAdd</a></li><li><a href="Photostream.html">Photostream</a></li><li><a href="SearchGroup.html">SearchGroup</a></li><li><a href="SearchPeople.html">SearchPeople</a></li><li><a href="SideNavBar.html">SideNavBar</a></li><li><a href="Upload.html">Upload</a></li><li><a href="UserInfo.html">UserInfo</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc/jsdoc">JSDoc 3.6.7</a> on Tue Jun 08 2021 19:30:23 GMT+0200 (Eastern European Standard Time)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>