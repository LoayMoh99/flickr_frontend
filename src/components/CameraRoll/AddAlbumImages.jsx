import React,{useState , useEffect } from 'react';
import axios from "axios"
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCheckCircle } from '@fortawesome/free-solid-svg-icons';
const endpoint = 'http://localhost:3001/'

function AddAlbumImages(props) {
    const check = <FontAwesomeIcon icon={faCheckCircle} color="cornflowerblue" size="1x"/>;
    const  { albumCover , albumTitle , albumPhotos , imgAddIds , albumId }= props;
    const [photoCount,setPhotoCount]= useState(albumPhotos.length);
    const [isChecked,setIsChecked]= useState(false);

    console.log(imgAddIds);
    console.log(albumId);
    
    const imgSelected={"photos":imgAddIds};

    async function deletePhotoFromAlbum (id,imgIdsToDelete)
    {
        try{
           // const response = await axios.delete(endpoint+'album/photos?album_id='+id, imgIdsToDelete) ;
            const response = await axios.delete(endpoint+'album/photos/'+id, imgIdsToDelete) ;
            setPhotoCount(response.data.length);
            return(response)
        } catch (error){
            if (error.response){
            /*
            * The request was made and the server responded with a
            * status code that falls out of the range of 2xx
            */
              console.log(error.response.data);
              console.log(error.response.status);
              console.log(error.response.headers);
            } else if (error.request){
              /*
              * The request was made but no response was received, `error.request`
              * is an instance of XMLHttpRequest in the browser and an instance
              * of http.ClientRequest in Node.js
              */
              console.log(error.request);
            } else {
              // Something happened in setting up the request and triggered an Error
              console.log('Error', error.message);
            }
            console.log(error);
        }
      };


      async function addPhotoToAlbum (id, object)
      {
                try{
             //const response = await axios.post(endpoint+'album?album_id='+id, object);
            const response = await axios.post(endpoint+'album/'+id , object);
            //Success
            setPhotoCount(response.data.length);
            return(response)
        } catch (error){
            if (error.response){
              console.log(error.response.data);
              console.log(error.response.status);
              console.log(error.response.headers);
            } else if (error.request){
              console.log(error.request);
            } else {
              console.log('Error', error.message);
            }
            console.log(error);
        }
      };
    


    // to toggle isChecked boolean
    const toggleChecked = () => {
        if(isChecked) // was initially checked, hence the user want to delete this/these photos from the selected album
        {
            deletePhotoFromAlbum (albumId,imgSelected);
        }
        else{ // was initially unchecked, hence the user want to add this/ these photos to the selected album 
            addPhotoToAlbum(albumId,imgSelected);
        }
        setIsChecked(!isChecked);
      };
  
      function containsPhotos(obj) {
        return imgAddIds.some((elem) => elem._id === obj._id);
      }
  
    // to check whether the selected photos ids are all present in a specific album or not
    const isPhotoPresent =()=>{
        // for loop on elements of selected photos ids 
        // compare the id in each loop  with the ids present in the album
        // if true (the id is present) then count ++
        // after looping .. check whether the count is equal to array of selected ids
        // if yes .. this means all selected ids are present in this album .. then make it checked
        // if no .. not all the selected photos ids are present .. then make it unchecked

        let count =0;
        albumPhotos.forEach(element => {
            if(containsPhotos(element))
            {
                count = count + 1 ;
            }
        }); 
        if(count === imgAddIds.length)
        {
            setIsChecked(true);
        }
        else
        {
            setIsChecked(false);
        }
        


    };
  
    useEffect(() => {
        isPhotoPresent()
    },[]);


    return (
      <>

            <div className="userAlbumAdd" onClick={toggleChecked}>
              <div className="userAlbumCover">
                {/* <div className="row"> */}
                  <img className="img-responsive" id="contentAdd" src={albumCover} alt="image_flickr" />
                  <div>
                    <h5 className="albumTitleAdd">{albumTitle}</h5>
                    <h5 className="albumInfoAdd">
                    {photoCount}
                    {' '}
                    items
                    </h5>
                  </div>
                  {/* <div> */}
                  {isChecked&&
                  (
                  <div className="checkIcon">
                    {check}
                  </div>
                  )}
                  {/* </div> */}
                {/* </div> */}
            </div>
            </div>

      </>
     );
    }
    
export default AddAlbumImages;
    