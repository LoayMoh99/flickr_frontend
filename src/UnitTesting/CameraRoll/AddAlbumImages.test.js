import React from 'react';
import ReactDOM from 'react-dom';
import AddAlbumImages from '../../components/CameraRoll/AddAlbumImages';
import Enzyme, { shallow, render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() })



it('AddAlbumImages renders without crashing', ()=>{
    const div = document.createElement('div');
    ReactDOM.render(<AddAlbumImages 
        albumCover = "https://picsum.photos/600/200?random=2"
        albumTitle = "Title"
        albumId = "1"
        albumPhotos = "[{_id:1},{_id:2}]"
        imgAddIds = "[1,2]"
       />, div);

});

it('AddAlbumImages renders correctly enzyme', () => {
    const wrapper = shallow(<AddAlbumImages 
        albumCover = "https://picsum.photos/600/200?random=2"
        albumTitle = "Title"
        albumId = "1"
        albumPhotos = "[{_id:1},{_id:2}]"
        imgAddIds = "[1,2]"
       />)
  
    expect(toJson(wrapper)).toMatchSnapshot();
  });