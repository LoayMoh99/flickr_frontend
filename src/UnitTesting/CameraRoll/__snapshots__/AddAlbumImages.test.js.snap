// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAlbumImages renders correctly enzyme 1`] = `
<Fragment>
  <div
    className="userAlbumAdd"
    onClick={[Function]}
  >
    <div
      className="userAlbumCover"
    >
      <img
        alt="image_flickr"
        className="img-responsive"
        id="contentAdd"
        src="https://picsum.photos/600/200?random=2"
      />
      <div
        className="albumAddInfo"
      >
        <h5
          className="albumTitleAdd"
        >
          Title
        </h5>
        <h5
          className="albumInfoAdd"
        >
          17
           
          items
        </h5>
      </div>
      <div
        className="checkIcon"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          color="cornflowerblue"
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f058",
                "M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z",
              ],
              "iconName": "check-circle",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size="1x"
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </div>
    </div>
  </div>
</Fragment>
`;
